main.c

```diff
@@ -0,0 +1,378 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#define _GNU_SOURCE /* needed for splice() */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <getopt.h>
+#include <mtd/ubi-user.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/sendfile.h>
+#include <sys/stat.h>
+#include <sys/sysmacros.h>
+#include <unistd.h>
+
+#include "config.h"
+
+#define UBIMajorDevNo 10
+#define BLOCK_SIZE (1024 * 1024L) /* 1 MiB */
+#define MIN(X, Y) ((X < Y) ? X : Y)
+
+static struct option long_options[] = {
+	{"help", no_argument, 0, 'h'},
+	{"write-everything", no_argument, 0, 'w'},
+	{"input-size", required_argument, 0, 's'},
+	{"fsync-interval", required_argument, 0, 'f'},
+	{"input", required_argument, 0, 'i'},
+	{"output", required_argument, 0, 'o'},
+	{0, 0, 0, 0}};
+
+void PrintHelp() {
+	fputs(
+		"Usage:\n"
+		"  mender-flash [-h|--help] [-w|--write-everything] [-s|--input-size <INPUT_SIZE>] [-f|--fsync-interval <FSYNC_INTERVAL>] -i|--input <INPUT_PATH> -o|--output <OUTPUT_PATH>\n",
+		stderr);
+}
+
+typedef ssize_t (*io_fn_t)(int, void *, size_t);
+
+struct Stats {
+	size_t blocks_written;
+	size_t blocks_omitted;
+	uint64_t bytes_written;
+	uint64_t bytes_omitted;
+	uint64_t total_bytes;
+};
+
+ssize_t buf_io(io_fn_t io_fn, int fd, unsigned char *buf, size_t len) {
+	size_t rem = len;
+	ssize_t n_done;
+	do {
+		n_done = io_fn(fd, buf + (len - rem), rem);
+		if (n_done > 0) {
+			rem -= n_done;
+		} else if ((n_done == -1) && (errno == EINTR)) {
```

Comment:
No, because `n_done` will be negative and thus the loop will stop on the condition check. The `continue` in this specific case prevents that from happening because it skips the loop condition check.

Reply:
But it triggered me to add one more test (see the 8th commit here). And now we are at
```
Overall coverage rate:
  lines......: 83.4% (156 of 187 lines)
  functions..: 100.0% (5 of 5 functions)
```
:partying_face: